napari/layers/utils/color_manager.py

```diff
@@ -0,0 +1,201 @@
+from dataclasses import field
+from itertools import cycle
+
+import numpy as np
+
+from ...utils.events.dataclass import Property, dataclass
+from ..utils.color_transformations import (
+    ColorType,
+    normalize_and_broadcast_colors,
+    transform_color,
+    transform_color_cycle,
+    transform_color_with_defaults,
+)
+from ._color_manager_constants import ColorMode
+from ._color_manager_utils import is_color_mapped
+from .layer_utils import guess_continuous
+
+
+@dataclass(events=True, properties=True)
+class ColorManager:
+    """Colors for a display property
+
+    Parameters
+    ----------
+    colors : np.ndarray
+        The RGBA color for each data entry
+    mode : ColorMode
+        Color setting mode. Should be one of the following:
+        DIRECT (default mode) allows each point to be set arbitrarily
+
+        CYCLE allows the color to be set via a color cycle over an attribute
+
+        COLORMAP allows color to be set via a color map over an attribute
+    color_cycle_map : dict
+        Mapping of categorical property values to colors
+
+    """
+
+    colors: np.ndarray = np.empty((0, 4))
+    mode: Property[ColorMode, str, None] = ColorMode.DIRECT
+    color_property: str = ''
+    color_cycle: Property[cycle, None] = cycle(
```

Comment:
haven't looked too carefully yet, but just fyi: we haven't yet tried to _both_ set a `dataclass` field _and_ use that same field name as a conventional `@property` (as is done with `color_cycle` here.  I think that might not be the way we want to do this... I'll need to spend a little more time with it to suggest an alternative though.

Reply:
Thanks for the feedback, @tlambert03 ! I am still wrapping my head around the dataclass, but I think I was coming to a similar conclusion. The alternative approach I am now working on is to create a `CategoricalColormap` class that holds all of the data/methods related to setting colors for categorical properties. I _think_ I can then remove this strange field+property pattern that I'm using here.